internal/pkg/scorecard/plugin_runner.go

```diff
@@ -0,0 +1,380 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scorecard
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/ghodss/yaml"
+	olmapiv1alpha1 "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	olminstall "github.com/operator-framework/operator-lifecycle-manager/pkg/controller/install"
+	k8sInternal "github.com/operator-framework/operator-sdk/internal/util/k8sutil"
+	"github.com/operator-framework/operator-sdk/internal/util/projutil"
+	"github.com/operator-framework/operator-sdk/internal/util/yamlutil"
+	scapiv1alpha1 "github.com/operator-framework/operator-sdk/pkg/apis/scorecard/v1alpha1"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	v1 "k8s.io/api/core/v1"
+	extscheme "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/scheme"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/discovery/cached"
+	"k8s.io/client-go/kubernetes"
+	cgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ConfigOpt                 = "config"
+	NamespaceOpt              = "namespace"
+	KubeconfigOpt             = "kubeconfig"
+	InitTimeoutOpt            = "init-timeout"
+	OlmDeployedOpt            = "olm-deployed"
+	CSVPathOpt                = "csv-path"
+	BasicTestsOpt             = "basic-tests"
+	OLMTestsOpt               = "olm-tests"
+	NamespacedManifestOpt     = "namespaced-manifest"
+	GlobalManifestOpt         = "global-manifest"
+	CRManifestOpt             = "cr-manifest"
+	ProxyImageOpt             = "proxy-image"
+	ProxyPullPolicyOpt        = "proxy-pull-policy"
+	CRDsDirOpt                = "crds-dir"
+	DeployDirOpt              = "deploy-dir"
+	OutputFormatOpt           = "output"
+	PluginDirOpt              = "plugin-dir"
+	JSONOutputFormat          = "json"
+	HumanReadableOutputFormat = "human-readable"
+)
+
+const (
+	basicOperator  = "Basic Operator"
+	olmIntegration = "OLM Integration"
+)
+
+var (
+	kubeconfig     *rest.Config
+	dynamicDecoder runtime.Decoder
+	runtimeClient  client.Client
+	restMapper     *restmapper.DeferredDiscoveryRESTMapper
+	deploymentName string
+	proxyPodGlobal *v1.Pod
+	cleanupFns     []cleanupFn
+)
+
+const (
+	scorecardPodName       = "operator-scorecard-test"
+	scorecardContainerName = "scorecard-proxy"
+)
+
+func SetupAndRunPlugin() (*scapiv1alpha1.ScorecardOutput, error) {
```

Comment:
I don't think there were any changes in the copy paste other than changing the logging so that the suite's logs get directed to the SuiteResult's log field and other logging goes to stderr (and thus the main scorecard log).

I think it would be useful to decouple the function from viper. I'm not sure if it would really make much sense to separate the basic and olm suites into separate functions though. They have a pretty much identical setup and teardown and the only difference are ~20 lines at line 287. Other than that, the setup that's done for the tests is the same, as is the teardown.

Reply:
> They have a pretty much identical setup and teardown and the only difference are ~20 lines at line 287. Other than that, the setup that's done for the tests is the same, as is the teardown.

Could we extract that common setup and teardown code into separate helper functions, such that the Basic and OLM functions end up being pretty simple?

```go
type CommonOpts {
    CommonOpt1 bool
}

type BasicOpts struct {
    CommonOpts

    BasicOpt1 string
}

type OLMOpts struct {
    CommonOpts

    OLMOpt1 string
}

func Basic(opts BasicOpts) (*scapiv1alpha1.ScorecardOutput, error) {
    if err := setup(opts.CommonOpts); err != nil {
        return nil, err
    }
    output, err := doBasicTests(opts)
    if err != nil {
        return nil, err
    }
    if err := teardown(opts.CommonOpts); err != nil {
        return nil, err
    }
    return output, nil
}

func OLM(opts OLMOpts) (*scapiv1alpha1.ScorecardOutput, error) {
    if err := setup(opts.CommonOpts); err != nil {
        return nil, err
    }
    output, err := doOLMTests(opts)
    if err != nil {
        return nil, err
    }
    if err := teardown(opts.CommonOpts); err != nil {
        return nil, err
    }
    return output, nil
}

Reply:
Yeah, that would be good to do in another PR. The current function is quite massive right now and should be modularized.